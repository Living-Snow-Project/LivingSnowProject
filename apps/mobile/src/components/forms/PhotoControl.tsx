import React, { useCallback } from "react";
import { Image, LogBox, Pressable, StyleSheet, Text, View } from "react-native";
import { SelectedPhoto } from "@livingsnow/record";
import { RootStackNavigationProp } from "../../navigation/Routes";
import { formInputStyles } from "../../styles/FormInput";
import TestIds from "../../constants/TestIds";
import { PictureIcon } from "../Icons";

// because we pass a callback in params, more info from the following links
// follow links for best practices, look at Context
// https://reactnavigation.org/docs/troubleshooting/#i-get-the-warning-non-serializable-values-were-found-in-the-navigation-state
// https://reactnavigation.org/docs/params/
// https://reactnavigation.org/docs/hello-react-navigation/#passing-additional-props
LogBox.ignoreLogs([
  "Non-serializable values were found in the navigation state",
]);

const styles = StyleSheet.create({
  // these styles could be dynamically sized\arranged based on number of photos and combinaions of portrait\landscape
  container: {
    flex: 1,
    margin: 1,
    height: 100,
  },
  photo: {
    width: "100%",
    height: "100%",
    borderColor: "black",
    borderWidth: 1,
    borderRadius: 2,
  },
});

type PhotoControlProps = {
  navigation: RootStackNavigationProp;
  photos: SelectedPhoto[];
  onUpdatePhotos: (photos: SelectedPhoto[]) => void;
};

export default function PhotoControl({
  navigation,
  photos,
  onUpdatePhotos,
}: PhotoControlProps) {
  const renderPhotos = useCallback(() => {
    const colsPerRow = 2;

    // layout algorithm for even number of photos
    if (photos.length % colsPerRow === 0) {
      const result: JSX.Element[] = [];
      const rows = photos.length / colsPerRow;

      for (let row = 0; row < rows; row++) {
        const inner: JSX.Element[] = [];
        for (let col = 0; col < colsPerRow; col++) {
          const index = row * rows + col;
          inner.push(
            <View style={styles.container} key={index}>
              <Image style={styles.photo} source={{ uri: photos[index].uri }} />
            </View>
          );
        }

        result.push(
          <View style={{ flex: rows, flexDirection: "row" }} key={row}>
            {inner}
          </View>
        );
      }

      // short circuit function
      return result;
    }

    // layout algorithm for odd number of photos
    return (
      <View style={{ flex: photos.length, flexDirection: "row" }}>
        {photos.map((x, index) => (
          /* eslint-disable react/no-array-index-key */
          <View style={styles.container} key={index}>
            <Image style={styles.photo} source={{ uri: x.uri }} />
          </View>
        ))}
      </View>
    );
  }, [photos]);

  return (
    <>
      <Text style={formInputStyles.optionStaticText}>
        Select Photos (limit 4)
      </Text>
      <Pressable
        testID={TestIds.Photos.photoSelectorTestId}
        onPress={() => {
          let existingSelection: string[] | undefined;

          if (photos.length > 0) {
            existingSelection = [];
            for (let i = 0; i < photos.length; i++) {
              existingSelection.push(photos[i].id);
            }
          }

          navigation.navigate("ImageSelection", {
            existingSelection,
            onUpdatePhotos,
          });
        }}
      >
        {photos.length === 0 && <PictureIcon />}
        {photos.length > 0 && renderPhotos()}
      </Pressable>
    </>
  );
}
