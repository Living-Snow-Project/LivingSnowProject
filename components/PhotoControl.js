import React from 'react';
import { Image, Pressable, View } from 'react-native';
import PropTypes from 'prop-types';
import { PictureIcon } from '../components/TabBarIcon';

const PhotoControl = ({navigation, photos}) => {
  return (
    <Pressable onPress={() => navigation.navigate(`Images`)}>
      {photos.length === 0 && <PictureIcon/>}
      {photos.length > 0 && 
      <View style={{flex:photos.length, flexDirection: "row"}}>{
        photos.map((x, index) =>
        <View style={{flex:1, margin:3}} key={index}>
          <Image style={{width: "100%", height: 100}} key={index} source={{uri: x.uri}}/>
        </View>)}
      </View>}
    </Pressable>
  );
}

PhotoControl.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired,
  }),
  photos: PropTypes.array,
};

export { PhotoControl };