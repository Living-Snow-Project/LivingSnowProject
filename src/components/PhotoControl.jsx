import React, { useCallback } from "react";
import { Image, LogBox, Pressable, StyleSheet, Text, View } from "react-native";
import PropTypes from "prop-types";
import { formInputStyles } from "../styles/FormInput";
import { PictureIcon } from "./TabBarIcon";
import Routes from "../navigation/Routes";

// because we pass a callback in params, more info from the following links
// follow links for best practices, look at Context
// https://reactnavigation.org/docs/troubleshooting/#i-get-the-warning-non-serializable-values-were-found-in-the-navigation-state
// https://reactnavigation.org/docs/params/
// https://reactnavigation.org/docs/hello-react-navigation/#passing-additional-props
LogBox.ignoreLogs([
  "Non-serializable values were found in the navigation state",
]);

export default function PhotoControl({ navigation, photos, onUpdatePhotos }) {
  const renderPhotos = useCallback(() => {
    const colsPerRow = 2;

    // layout algorithm for even number of photos
    if (photos.length % colsPerRow === 0) {
      const result = [];
      const rows = photos.length / colsPerRow;

      for (let row = 0; row < rows; row++) {
        const inner = [];
        for (let col = 0; col < colsPerRow; col++) {
          const index = row * rows + col;
          inner.push(
            <View style={styles.singlePhotoContainer} key={index}>
              <Image style={styles.photo} source={{ uri: photos[index].uri }} />
            </View>
          );
        }

        result.push(
          <View style={{ flex: rows, flexDirection: "row" }} key={row}>
            {inner}
          </View>
        );
      }

      // short circuit function
      return result;
    }

    // layout algorithm for odd number of photos
    return (
      <View style={{ flex: photos.length, flexDirection: "row" }}>
        {photos.map((x, index) => (
          <View style={styles.singlePhotoContainer} key={index}>
            <Image style={styles.photo} key={index} source={{ uri: x.uri }} />
          </View>
        ))}
      </View>
    );
  }, [photos]);

  return (
    <>
      <Text style={formInputStyles.optionStaticText}>
        Select Photos (limit 4)
      </Text>
      <Pressable
        onPress={() =>
          navigation.navigate(Routes.ImagesPickerScreen, {
            onUpdatePhotos,
          })
        }
      >
        {photos.length === 0 && <PictureIcon />}
        {photos.length > 0 && renderPhotos()}
      </Pressable>
    </>
  );
}

PhotoControl.propTypes = {
  photos: PropTypes.arrayOf(
    PropTypes.oneOfType([
      PropTypes.shape({
        height: PropTypes.number,
        uri: PropTypes.string,
        width: PropTypes.number,
      }),
    ])
  ),
  onUpdatePhotos: PropTypes.func.isRequired,
};

PhotoControl.defaultProps = {
  photos: undefined,
};

const styles = StyleSheet.create({
  // these styles could be dynamically sized\arranged based on number of photos and combinaions of portrait\landscape
  singlePhotoContainer: {
    flex: 1,
    margin: 1,
    height: 100,
  },
  photo: {
    width: "100%",
    height: "100%",
    borderColor: "black",
    borderWidth: 1,
    borderRadius: 2,
  },
});
