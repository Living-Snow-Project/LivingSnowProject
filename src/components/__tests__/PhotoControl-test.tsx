import React from "react";
import { render } from "@testing-library/react-native";
import PhotoControl from "../PhotoControl";

const navigation = {
  navigate: () => {},
  goBack: () => {},
};

describe("PhotoControl test suite", () => {
  test("even number of photos", () => {
    const photos = [
      { uri: "", height: 16, width: 24 },
      { uri: "", height: 24, width: 16 },
    ];

    const { toJSON } = render(
      <PhotoControl
        navigation={navigation}
        photos={photos}
        onUpdatePhotos={() => {}}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  test("odd number of photos", () => {
    const photos = [{ uri: "", height: 16, width: 24 }];

    const { toJSON } = render(
      <PhotoControl
        navigation={navigation}
        photos={photos}
        onUpdatePhotos={() => {}}
      />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
