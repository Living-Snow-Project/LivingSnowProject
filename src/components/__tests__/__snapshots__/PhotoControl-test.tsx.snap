// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoControl test suite even number of photos 1`] = `
Array [
  <Text
    style={
      Object {
        "fontSize": 15,
        "marginTop": 3,
      }
    }
  >
    Select Photos (limit 4)
  </Text>,
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="photo-selector"
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "height": 100,
            "margin": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "borderColor": "black",
              "borderRadius": 2,
              "borderWidth": 1,
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "height": 100,
            "margin": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "borderColor": "black",
              "borderRadius": 2,
              "borderWidth": 1,
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </View>
    </View>
  </View>,
]
`;

exports[`PhotoControl test suite odd number of photos 1`] = `
Array [
  <Text
    style={
      Object {
        "fontSize": 15,
        "marginTop": 3,
      }
    }
  >
    Select Photos (limit 4)
  </Text>,
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="photo-selector"
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "height": 100,
            "margin": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "borderColor": "black",
              "borderRadius": 2,
              "borderWidth": 1,
              "height": "100%",
              "width": "100%",
            }
          }
        />
      </View>
    </View>
  </View>,
]
`;
